phoenix-core/src/main/java/org/apache/phoenix/monitoring/TableMetricsManager.java

```diff
@@ -0,0 +1,287 @@
+ */
+
+package org.apache.phoenix.monitoring;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import org.apache.hadoop.hbase.util.EnvironmentEdgeManager;
+import org.apache.phoenix.query.QueryServicesOptions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Central place where we keep track of all the Table Level metrics. Register each tableMetrics and
+ * store the instance of it associated with TableName in a map
+ * This class exposes following functions as static methods to help catch all execptions
+ * 1.clearTableLevelMetricsMethod
+ * 2.getTableMetricsMethod
+ * 3.pushMetricsFromConnInstanceMethod
+ * 4.updateMetricsMethod
+ */
+
+public class TableMetricsManager {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TableMetricsManager.class);
+    private static boolean isTableLevelMetricsEnabled;
+    private static boolean isMetricPublisherEnabled;
+    private static final Set<String> allowedListOfTableNames = new HashSet<>();
+    private static ConcurrentMap<String, TableClientMetrics> tableClientMetricsMapping = null;
+    // Singleton object
+    protected static TableMetricsManager tableMetricsManager = null;
+    private static MetricPublisherSupplierFactory mPublisher = null;
+    private static QueryServicesOptions options = null;
+
+    public TableMetricsManager(QueryServicesOptions ops) {
+        options = ops;
+        isTableLevelMetricsEnabled = options.isTableLevelMetricsEnabled();
+        LOGGER.info(String.format("Phoenix Table metrics enabled status: %s",
+                isTableLevelMetricsEnabled));
+        tableClientMetricsMapping = new ConcurrentHashMap<>();
+
+        String tableNamesList = options.getAllowedListTableNames();
+        if (tableNamesList != null && !tableNamesList.isEmpty()) {
+            for (String tableName : tableNamesList.split(",")) {
+                allowedListOfTableNames.add(tableName);
+            }
+        }
+        isMetricPublisherEnabled = options.isMetricPublisherEnabled();
+        LOGGER.info(String.format("Phoenix table level metrics publisher enabled status %s",
+                isMetricPublisherEnabled));
+    }
+
+    public TableMetricsManager() {
+    }
+
+    @VisibleForTesting public static void setInstance(TableMetricsManager metricsManager) {
+        tableMetricsManager = metricsManager;
+    }
+
+    /**
+     * Method to provide instance of TableMetricsManager(Create if needed in thread safe manner)
+     *
+     * @return
+     */
+    private static TableMetricsManager getInstance() {
+
+        if (tableMetricsManager == null) {
+            synchronized (TableMetricsManager.class) {
+                if (tableMetricsManager == null) {
```

Comment:
For double-checked locking to work properly, `tableMetricsManager` will need to be declared `volatile`.

Reply:
Same with other variables of this class, need to be either `volatile` or if possible `final`

Reply:
Also, consider using a local variable to point to the volatile field for performance reasons (See https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java for more details)

Reply:
Thanks, Modified the implementation.